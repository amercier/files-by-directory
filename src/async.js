import regeneratorRuntime from 'regenerator-runtime';

/**
 * Naive Promisify implementation.
 *
 * Implementing this instead of using NodeJS `util.Promisify` allow to maintain NodeJS 6 support.
 * @see {@link https://nodejs.org/dist/latest-v8.x/docs/api/util.html#util_util_promisify_original|util.promisify(original)}
 *
 * @param {Function} fn Function of the form: (err, callback).
 * @returns {Function} A function that returns a Promise instead.
 */
export function promisify(fn) {
  return (...args) =>
    new Promise((resolve, reject) =>
      fn(...args, (err, result) => (err ? reject(err) : resolve(result))),
    );
}

/**
 * Apply a given function to each elements of a given iterable, and yield each returned value.
 *
 * @async
 * @generator
 * @param {Iterable<T>} iterable Synchronous or asynchronous iterable.
 * @param {Function: T => U} fn Synchronous or asynchronous function.
 * @returns {AsyncIterator<U>} An asynchronous iterator that yields value returned by `fn` applied to
 * each element of `iterable`.
 */
export async function* asyncMap(iterable, fn) {
  for await (const value of iterable) {
    yield fn(value);
  }
}

/**
 * Return an asynchronous generator that yields all elements from a given iterable that pass the
 * test implemented by the provided function.
 *
 * @async
 * @generator
 * @param {Iterable<T>} iterable Synchronous or asynchronous iterable.
 * @param {Function: T => boolean} filter Synchronous or asynchronous function.
 * @returns {AsyncIterator<T>} An asynchronous iterator that yields values from `iterable` where
 * `filter` return a truthy value.
 */
export async function* asyncFilter(iterable, filter) {
  for await (const value of iterable) {
    if (await filter(value)) {
      yield value;
    }
  }
}

/**
 * Apply a given generator to each element of a given iterable, and re-yield every yielded values.
 *
 * @async
 * @generator
 * @param {Iterable<T>} iterable Synchronous or asynchronous iterable.
 * @param {Function: T => Iterable<U>} generator Synchronous or asynchronous generator function.
 * @returns {AsyncIterator<U>} An asynchronous iterator that re-yields each value yielded by `generator(iterable[i])`.
 */
export async function* asyncFlattenMap(iterable, generator) {
  for await (const value of iterable) {
    yield* generator(value);
  }
}

/**
 * Get all values from an iterable.
 *
 * @param {Iterable<T>} iterable Synchronous or asynchronous iterable.
 * @returns {Promise<T[]>} A Promise resolving to an array containing all yielded values.
 */
export async function values(iterable) {
  const array = [];
  for await (const value of iterable) {
    array.push(value);
  }
  return array;
}
