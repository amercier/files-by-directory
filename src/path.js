import escape from 'escape-string-regexp';
import { dirname, relative, sep } from 'path';

/**
 * Group paths by their parent path.
 *
 * @param {string[]} paths Paths.
 * @returns {Object} Plain object indexed by parent paths containing paths arrays.
 */
export function byDirname(paths) {
  return paths.reduce((acc, file) => {
    const parent = dirname(file.path);
    if (!acc[parent]) {
      acc[parent] = [];
    }
    acc[parent].push(file);
    return acc;
  }, {});
}

/**
 * Regular expression that matches all "ascendant" paths:
 * - '..'
 * - '../..'
 * - '../../..'
 * - etc.
 *
 * @type {RegExp}
 */
export const ascendanceRegExp = new RegExp(`^(${escape('..')}${escape(sep)})*${escape('..')}$`);

/**
 * Whether a path is an ascendant of another one.
 *
 * @param {string} subject ascendant path.
 * @param {string} descendant descendant path.
 * @returns {boolean} `true` if `subject` is an ascendant of `descendant`, `false` otherwise.
 */
export function isAscendant(subject, descendant) {
  return ascendanceRegExp.test(relative(descendant, subject));
}

/**
 * Whether a path is an descendant of another one.
 *
 * @param {string} subject descendant path.
 * @param {string} ascendant ascendant path.
 * @returns {boolean} `true` if `subject` is an ascendant of `ascendant`, `false` otherwise.
 */
export function isDescendant(subject, ascendant) {
  return isAscendant(ascendant, subject);
}

/**
 * Whether a path is unique and does not have any ascendant within an array of paths.
 *
 * @param {string} path A path.
 * @param {number} index Index of `path` with
 * @param {string[]} paths Array of paths.
 * @returns {boolean} `true` if `path` is both unique and do not have any ascendant within `paths`,
 * `false` otherwise.
 */
export function isUniqueAndNotDescendant(path, index, paths) {
  return paths.indexOf(path) === index && !paths.some(p => isAscendant(p, path));
}
