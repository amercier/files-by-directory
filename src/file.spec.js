import '@babel/polyfill'; // Required for NodeJS < 10
import { basename, dirname } from 'path';
import { values } from './async';
import File from './file';
import {
  file1a,
  level1,
  level2,
  level3,
  linkToSiblingDirectory,
  linkToSiblingFile,
  linkToUnexistingFile,
  unexistingFile,
} from '../fixture';

const file1aArgs = [file1a, false, false];
const level1Args = [level1, true, false];
const level2Args = [level2, true, false];
const level3Args = [level3, true, false];
const linkToSiblingDirectoryArgs = [linkToSiblingDirectory, false, true];
const linkToSiblingFileArgs = [linkToSiblingFile, false, true];
const linkToUnexistingFileArgs = [linkToUnexistingFile, false, true];
const unexistingFileArgs = [unexistingFile, false, false];

describe('File', () => {
  describe('constructor()', () => {
    it('creates instances of File', () => {
      expect(new File(...file1aArgs)).toBeInstanceOf(File);
      expect(new File(...level1Args)).toBeInstanceOf(File);
      expect(new File(...level2Args)).toBeInstanceOf(File);
      expect(new File(...level3Args)).toBeInstanceOf(File);
      expect(new File(...unexistingFileArgs)).toBeInstanceOf(File);
      expect(new File(...linkToSiblingDirectoryArgs)).toBeInstanceOf(File);
      expect(new File(...linkToSiblingFileArgs)).toBeInstanceOf(File);
      expect(new File(...linkToUnexistingFileArgs)).toBeInstanceOf(File);
    });

    it('stores path', () => {
      expect(new File(...file1aArgs).path).toBe(file1aArgs[0]);
      expect(new File(...level1Args).path).toBe(level1Args[0]);
      expect(new File(...level2Args).path).toBe(level2Args[0]);
      expect(new File(...level3Args).path).toBe(level3Args[0]);
      expect(new File(...unexistingFileArgs).path).toBe(unexistingFileArgs[0]);
      expect(new File(...linkToSiblingDirectoryArgs).path).toBe(linkToSiblingDirectoryArgs[0]);
      expect(new File(...linkToSiblingFileArgs).path).toBe(linkToSiblingFileArgs[0]);
      expect(new File(...linkToUnexistingFileArgs).path).toBe(linkToUnexistingFileArgs[0]);
    });

    it('stores isDirectory', () => {
      expect(new File(...file1aArgs).isDirectory).toBe(file1aArgs[1]);
      expect(new File(...level1Args).isDirectory).toBe(level1Args[1]);
      expect(new File(...level2Args).isDirectory).toBe(level2Args[1]);
      expect(new File(...level3Args).isDirectory).toBe(level3Args[1]);
      expect(new File(...unexistingFileArgs).isDirectory).toBe(unexistingFileArgs[1]);
      expect(new File(...linkToSiblingDirectoryArgs).isDirectory).toBe(
        linkToSiblingDirectoryArgs[1],
      );
      expect(new File(...linkToSiblingFileArgs).isDirectory).toBe(linkToSiblingFileArgs[1]);
      expect(new File(...linkToUnexistingFileArgs).isDirectory).toBe(linkToUnexistingFileArgs[1]);
    });
  });

  describe('getChildren()', () => {
    it('returns an iterator', () => {
      const iterator = new File(file1a, false).getChildren();
      expect(iterator.next).toBeFunction();
    });

    it('generates a rejected Promise when file is a file', async () => {
      await expect(values(new File(...file1aArgs).getChildren())).toReject();
    });

    it('generates asynchronously one File instance per child when file is a directory', async () => {
      expect(await values(new File(...level1Args).getChildren())).toMatchSnapshot();
      expect(await values(new File(...level2Args).getChildren())).toMatchSnapshot();
      expect(await values(new File(...level3Args).getChildren())).toMatchSnapshot();
    });

    it('generates a rejected Promise when file does not exist', async () => {
      await expect(values(new File(...unexistingFileArgs).getChildren())).toReject();
    });

    it('generates a rejected Promise when file links to a file', async () => {
      await expect(values(new File(...linkToSiblingFile).getChildren())).toReject();
    });

    it('generates a rejected Promise when file links to a directory', async () => {
      await expect(values(new File(...linkToSiblingDirectoryArgs).getChildren())).toReject();
    });

    it('generates a rejected Promise when file links to a non-existing file', async () => {
      await expect(values(new File(...linkToUnexistingFile).getChildren())).toReject();
    });
  });

  describe('static fromDirent()', () => {
    const argsToDirent = ([path, isDirectory, isSymbolicLink]) => ({
      name: basename(path),
      isDirectory: () => isDirectory,
      isSymbolicLink: () => isSymbolicLink,
    });
    const file1aDirent = argsToDirent(file1aArgs);
    const level2Dirent = argsToDirent(level2Args);
    const level3Dirent = argsToDirent(level3Args);
    const unexistingFileDirent = argsToDirent(unexistingFileArgs);
    const linkToSiblingDirectoryDirent = argsToDirent(linkToSiblingDirectoryArgs);
    const linkToSiblingFileDirent = argsToDirent(linkToSiblingFileArgs);
    const linkToUnexistingFileDirent = argsToDirent(linkToUnexistingFileArgs);

    it('creates instances of File', () => {
      expect(File.fromDirent(level1, file1aDirent)).toBeInstanceOf(File);
    });

    it('it passes given path and name', () => {
      [
        [file1a, file1aDirent],
        [level2, level2Dirent],
        [level3, level3Dirent],
        [unexistingFile, unexistingFileDirent],
        [linkToSiblingFile, linkToSiblingFileDirent],
        [linkToSiblingDirectory, linkToSiblingDirectoryDirent],
        [linkToUnexistingFile, linkToUnexistingFileDirent],
      ].forEach(([path, dirent]) => {
        expect(File.fromDirent(dirname(path), dirent).path).toBe(path);
      });
    });

    it('it uses given dirent to determine isDirectory', () => {
      [
        [file1a, file1aDirent, file1aArgs[1]],
        [level2, level2Dirent, level2Args[1]],
        [level3, level3Dirent, level3Args[1]],
        [unexistingFile, unexistingFileDirent, unexistingFileArgs[1]],
        [linkToSiblingFile, linkToSiblingFileDirent, linkToSiblingFileArgs[1]],
        [linkToSiblingDirectory, linkToSiblingDirectoryDirent, linkToSiblingDirectoryArgs[1]],
        [linkToUnexistingFile, linkToUnexistingFileDirent, linkToUnexistingFileArgs[1]],
      ].forEach(([path, dirent, isDirectory]) => {
        expect(File.fromDirent(dirname(path), dirent).isDirectory).toBe(isDirectory);
      });
    });

    it('it uses given dirent to determine isSymbolicLink', () => {
      [
        [file1a, file1aDirent, file1aArgs[2]],
        [level2, level2Dirent, level2Args[2]],
        [level3, level3Dirent, level3Args[2]],
        [unexistingFile, unexistingFileDirent, unexistingFileArgs[2]],
        [linkToSiblingFile, linkToSiblingFileDirent, linkToSiblingFileArgs[2]],
        [linkToSiblingDirectory, linkToSiblingDirectoryDirent, linkToSiblingDirectoryArgs[2]],
        [linkToUnexistingFile, linkToUnexistingFileDirent, linkToUnexistingFileArgs[2]],
      ].forEach(([path, dirent, isSymbolicLink]) => {
        expect(File.fromDirent(dirname(path), dirent).isSymbolicLink).toBe(isSymbolicLink);
      });
    });
  });

  describe('static fromPath()', () => {
    it('returns a Promise', () => {
      expect(File.fromPath(file1a)).toBeInstanceOf(Promise);
    });

    it('resolves with an instance of File when file is a directory', async () => {
      await expect(File.fromPath(level2)).toResolve();
      expect(await File.fromPath(level2)).toMatchSnapshot();
    });

    it('resolves with an instance of File when file is a file', async () => {
      await expect(File.fromPath(file1a)).toResolve();
      expect(await File.fromPath(file1a)).toMatchSnapshot();
    });

    it('rejects when file is an non-existing file', async () => {
      await expect(File.fromPath(unexistingFile)).toReject();
    });

    it('resolves with and instance of File when symlink links to a file', async () => {
      await expect(File.fromPath(linkToSiblingFile)).toResolve();
      expect(await File.fromPath(linkToSiblingFile)).toMatchSnapshot();
    });

    it('resolves with and instance of File when symlink links to a directory', async () => {
      await expect(File.fromPath(linkToSiblingDirectory)).toResolve();
      expect(await File.fromPath(linkToSiblingDirectory)).toMatchSnapshot();
    });

    it('resolves with and instance of File when symlink links to non-existing file', async () => {
      await expect(File.fromPath(linkToUnexistingFile)).toResolve();
      expect(await File.fromPath(linkToUnexistingFile)).toMatchSnapshot();
    });
  });

  describe('static fromPaths()', () => {
    it('generates instances of File asynchronously', async () => {
      for await (const file of File.fromPaths([
        file1a,
        level2,
        linkToSiblingFile,
        linkToSiblingDirectory,
        linkToUnexistingFile,
      ])) {
        expect(file).toMatchSnapshot();
      }
    });

    it('generates a rejected Promise when an non-existing file is given', async () => {
      const iterator = File.fromPaths([unexistingFile]);
      await expect(iterator.next()).toReject();
    });
  });

  describe('static readDir()', () => {
    it('generates a rejected Promise when path is a file', async () => {
      await expect(values(File.readDir(file1a))).toReject();
    });

    it('generates asynchronously one File instance per child when path is a directory', async () => {
      expect(await values(File.readDir(level3))).toMatchSnapshot();
      expect(await values(File.readDir(level2))).toMatchSnapshot();
      expect(await values(File.readDir(level1))).toMatchSnapshot();
    });

    it('generates a rejected Promise when path is a non-existing file', async () => {
      await expect(values(File.readDir(unexistingFile))).toReject();
    });

    it('generates a rejected Promise when path links to a file', async () => {
      await expect(values(File.readDir(linkToSiblingFile))).toReject();
    });

    it('generates asynchronously one File instance per child when path links to a directory', async () => {
      expect(await values(File.readDir(linkToSiblingDirectory))).toMatchSnapshot();
    });

    it('generates a rejected Promise when path links to a non-existing file', async () => {
      await expect(values(File.readDir(linkToUnexistingFile))).toReject();
    });
  });
});
