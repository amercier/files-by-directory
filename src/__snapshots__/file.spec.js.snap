// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File followSymbolicLink() resolves to a new File instance when file links to a directory 1`] = `
File {
  "isDirectory": true,
  "isSymbolicLink": false,
  "path": "fixture/level1/level2/link-to-sibling-directory",
}
`;

exports[`File followSymbolicLink() resolves to a new File instance when file links to a file 1`] = `
File {
  "isDirectory": false,
  "isSymbolicLink": false,
  "path": "fixture/level1/level2/link-to-sibling-file",
}
`;

exports[`File getChildren() generates asynchronously one File instance per child when file is a directory 1`] = `
Array [
  File {
    "isDirectory": false,
    "isSymbolicLink": false,
    "path": "fixture/level1/file1a",
  },
  File {
    "isDirectory": false,
    "isSymbolicLink": false,
    "path": "fixture/level1/file1b",
  },
  File {
    "isDirectory": true,
    "isSymbolicLink": false,
    "path": "fixture/level1/level2",
  },
  File {
    "isDirectory": undefined,
    "isSymbolicLink": true,
    "path": "fixture/level1/link-to-current-directory",
  },
  File {
    "isDirectory": undefined,
    "isSymbolicLink": true,
    "path": "fixture/level1/link-to-descendant-directory",
  },
  File {
    "isDirectory": undefined,
    "isSymbolicLink": true,
    "path": "fixture/level1/link-to-descendant-file",
  },
  File {
    "isDirectory": undefined,
    "isSymbolicLink": true,
    "path": "fixture/level1/link-to-grand-parent-directory",
  },
  File {
    "isDirectory": undefined,
    "isSymbolicLink": true,
    "path": "fixture/level1/link-to-parent-directory",
  },
  File {
    "isDirectory": undefined,
    "isSymbolicLink": true,
    "path": "fixture/level1/link-to-unexisting-file",
  },
]
`;

exports[`File getChildren() generates asynchronously one File instance per child when file is a directory 2`] = `
Array [
  File {
    "isDirectory": false,
    "isSymbolicLink": false,
    "path": "fixture/level1/level2/file2a",
  },
  File {
    "isDirectory": false,
    "isSymbolicLink": false,
    "path": "fixture/level1/level2/file2b",
  },
  File {
    "isDirectory": true,
    "isSymbolicLink": false,
    "path": "fixture/level1/level2/level3",
  },
  File {
    "isDirectory": undefined,
    "isSymbolicLink": true,
    "path": "fixture/level1/level2/link-to-sibling-directory",
  },
  File {
    "isDirectory": undefined,
    "isSymbolicLink": true,
    "path": "fixture/level1/level2/link-to-sibling-file",
  },
]
`;

exports[`File getChildren() generates asynchronously one File instance per child when file is a directory 3`] = `
Array [
  File {
    "isDirectory": false,
    "isSymbolicLink": false,
    "path": "fixture/level1/level2/level3/file3a",
  },
  File {
    "isDirectory": false,
    "isSymbolicLink": false,
    "path": "fixture/level1/level2/level3/file3b",
  },
]
`;

exports[`File static fromPath() resolves with an instance of File when file is a directory 1`] = `
File {
  "isDirectory": true,
  "isSymbolicLink": false,
  "path": "fixture/level1/level2",
}
`;

exports[`File static fromPath() resolves with an instance of File when file is a file 1`] = `
File {
  "isDirectory": false,
  "isSymbolicLink": false,
  "path": "fixture/level1/file1a",
}
`;

exports[`File static fromPath() resolves with and instance of File when symlink links to a directory 1`] = `
File {
  "isDirectory": undefined,
  "isSymbolicLink": true,
  "path": "fixture/level1/level2/link-to-sibling-directory",
}
`;

exports[`File static fromPath() resolves with and instance of File when symlink links to a file 1`] = `
File {
  "isDirectory": undefined,
  "isSymbolicLink": true,
  "path": "fixture/level1/level2/link-to-sibling-file",
}
`;

exports[`File static fromPath() resolves with and instance of File when symlink links to non-existing file 1`] = `
File {
  "isDirectory": undefined,
  "isSymbolicLink": true,
  "path": "fixture/level1/link-to-unexisting-file",
}
`;

exports[`File static fromPaths() generates instances of File asynchronously 1`] = `
File {
  "isDirectory": false,
  "isSymbolicLink": false,
  "path": "fixture/level1/file1a",
}
`;

exports[`File static fromPaths() generates instances of File asynchronously 2`] = `
File {
  "isDirectory": true,
  "isSymbolicLink": false,
  "path": "fixture/level1/level2",
}
`;

exports[`File static fromPaths() generates instances of File asynchronously 3`] = `
File {
  "isDirectory": undefined,
  "isSymbolicLink": true,
  "path": "fixture/level1/level2/link-to-sibling-file",
}
`;

exports[`File static fromPaths() generates instances of File asynchronously 4`] = `
File {
  "isDirectory": undefined,
  "isSymbolicLink": true,
  "path": "fixture/level1/level2/link-to-sibling-directory",
}
`;

exports[`File static fromPaths() generates instances of File asynchronously 5`] = `
File {
  "isDirectory": undefined,
  "isSymbolicLink": true,
  "path": "fixture/level1/link-to-unexisting-file",
}
`;

exports[`File static readDir() generates asynchronously one File instance per child when path is a directory 1`] = `
Array [
  File {
    "isDirectory": false,
    "isSymbolicLink": false,
    "path": "fixture/level1/level2/level3/file3a",
  },
  File {
    "isDirectory": false,
    "isSymbolicLink": false,
    "path": "fixture/level1/level2/level3/file3b",
  },
]
`;

exports[`File static readDir() generates asynchronously one File instance per child when path is a directory 2`] = `
Array [
  File {
    "isDirectory": false,
    "isSymbolicLink": false,
    "path": "fixture/level1/level2/file2a",
  },
  File {
    "isDirectory": false,
    "isSymbolicLink": false,
    "path": "fixture/level1/level2/file2b",
  },
  File {
    "isDirectory": true,
    "isSymbolicLink": false,
    "path": "fixture/level1/level2/level3",
  },
  File {
    "isDirectory": undefined,
    "isSymbolicLink": true,
    "path": "fixture/level1/level2/link-to-sibling-directory",
  },
  File {
    "isDirectory": undefined,
    "isSymbolicLink": true,
    "path": "fixture/level1/level2/link-to-sibling-file",
  },
]
`;

exports[`File static readDir() generates asynchronously one File instance per child when path is a directory 3`] = `
Array [
  File {
    "isDirectory": false,
    "isSymbolicLink": false,
    "path": "fixture/level1/file1a",
  },
  File {
    "isDirectory": false,
    "isSymbolicLink": false,
    "path": "fixture/level1/file1b",
  },
  File {
    "isDirectory": true,
    "isSymbolicLink": false,
    "path": "fixture/level1/level2",
  },
  File {
    "isDirectory": undefined,
    "isSymbolicLink": true,
    "path": "fixture/level1/link-to-current-directory",
  },
  File {
    "isDirectory": undefined,
    "isSymbolicLink": true,
    "path": "fixture/level1/link-to-descendant-directory",
  },
  File {
    "isDirectory": undefined,
    "isSymbolicLink": true,
    "path": "fixture/level1/link-to-descendant-file",
  },
  File {
    "isDirectory": undefined,
    "isSymbolicLink": true,
    "path": "fixture/level1/link-to-grand-parent-directory",
  },
  File {
    "isDirectory": undefined,
    "isSymbolicLink": true,
    "path": "fixture/level1/link-to-parent-directory",
  },
  File {
    "isDirectory": undefined,
    "isSymbolicLink": true,
    "path": "fixture/level1/link-to-unexisting-file",
  },
]
`;

exports[`File static readDir() generates asynchronously one File instance per child when path links to a directory 1`] = `
Array [
  File {
    "isDirectory": false,
    "isSymbolicLink": false,
    "path": "fixture/level1/level2/link-to-sibling-directory/file3a",
  },
  File {
    "isDirectory": false,
    "isSymbolicLink": false,
    "path": "fixture/level1/level2/link-to-sibling-directory/file3b",
  },
]
`;
